okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java

```diff
@@ -0,0 +1,138 @@
+ * <p>
+ * Concurrency permits are only released when the response body is closed.
+ */
+final class ConcurrencyLimitingInterceptor implements Interceptor {
+    private static final ImmutableSet<Integer> DROPPED_CODES = ImmutableSet.of(429, 503);
+
+    private final ConcurrencyLimiters limiters;
+
+    @VisibleForTesting
+    ConcurrencyLimitingInterceptor(ConcurrencyLimiters limiters) {
+        this.limiters = limiters;
+    }
+
+    ConcurrencyLimitingInterceptor() {
+        this(new ConcurrencyLimiters());
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Limiter.Listener listener = limiters.limiter(chain.request());
+        try {
+            Response response = chain.proceed(chain.request());
+            if (DROPPED_CODES.contains(response.code())) {
+                listener.onDropped();
+                return response;
+            } else if (!response.isSuccessful() || response.isRedirect()) {
+                listener.onIgnore();
+                return response;
+            } else {
+                return wrapResponse(listener, response);
+            }
+        } catch (IOException e) {
+            listener.onIgnore();
+            throw e;
+        }
+    }
+
+    private static Response wrapResponse(Limiter.Listener listener, Response response) {
+        if (response.body() == null) {
+            return response;
+        }
+        ResponseBody currentBody = response.body();
+        ResponseBody newResponseBody =
+                ResponseBody.create(
+                        currentBody.contentType(),
+                        currentBody.contentLength(),
+                        wrapSource(currentBody.source(), listener));
+        return response.newBuilder()
+                .body(newResponseBody)
+                .build();
+    }
+
+    private static BufferedSource wrapSource(BufferedSource currentSource, Limiter.Listener listener) {
+        return (BufferedSource) Proxy.newProxyInstance(
+                BufferedSource.class.getClassLoader(),
+                new Class<?>[] { BufferedSource.class },
+                new ReleaseConcurrencyLimitProxy(currentSource, listener));
+    }
+
+    /**
+     * This proxy enables e.g. Okio to make additive additions to their API without breaking us.
+     */
+    private static final class ReleaseConcurrencyLimitProxy implements InvocationHandler {
+        private final BufferedSource delegate;
+        private final Limiter.Listener listener;
+        private boolean closed = false;
+
+        private ReleaseConcurrencyLimitProxy(BufferedSource delegate, Limiter.Listener listener) {
+            this.delegate = delegate;
+            this.listener = listener;
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            if (method.getName().equals("close") && !closed) {
+                closed = true;
+                listener.onSuccess();
```

Comment:
So the closing mechanism is now implicit, in that as long as the response body is closed (either by reading it fully during json object mapping, or via `inputstream.close()` if streaming and want to finish earlier) we tag as successful?

Is it viable at some point to have more control over this? One thing that occasionally happens when streaming is your stream ends too early because an error was encountered once some data (esp. headers) was already sent. We would preferably mark those as failed.